session/store.go

```diff
@@ -13,22 +14,59 @@ var (
 	// sessionBucketKey is the top level bucket where we can find all
 	// information about sessions. These sessions are indexed by their
 	// public key.
+	//
+	// The session bucket has the following structure:
+	// session -> <key>       -> <serialised session>
+	//	   -> id-index    -> <session-id> -> key   -> <session key>
+	// 			                  -> group -> <group-ID>
+	// 	   -> linked-session-index -> <linked-ID> -> <ID>
+	// 	   -> group-id-index -> <group-id> -> session-id -> sequence -> <session-id>
 	sessionBucketKey = []byte("session")
 
+	// idIndexKey is the key used to define the id-index sub-bucket within
+	// the main session bucket. This bucket will be used to store the
+	// mapping from session ID to various other fields.
+	idIndexKey = []byte("id-index")
+
+	// sessionKeyKey is the key used within the id-index bucket to store the
+	// session key (serialised local public key) associated with the given
+	// session ID.
+	sessionKeyKey = []byte("key")
+
+	// linkedSessionIndex is an index from an older session to a newer
+	// session that is linked to it. A session can only ever be linked to
+	// by one other session.
+	linkedSessionIndex = []byte("linked-session-index")
+
+	// groupIDKey is the key used within the id-index bucket to store the
+	// group ID associated with the given session ID.
+	groupIDKey = []byte("group")
+
+	// groupIDIndexKey is the key used to define the group-id-index
+	// sub-bucket within the main session bucket. This bucket will be used
+	// to store the mapping from group ID to various other fields.
+	groupIDIndexKey = []byte("group-id-index")
+
+	// sessionIDKey is a key used in the group-id-index under a sub-bucket
+	// defined by a specific group ID. It will be used to store the session
+	// IDs associated with the given group ID.
+	sessionIDKey = []byte("session-id")
+
 	// ErrSessionNotFound is an error returned when we attempt to retrieve
 	// information about a session but it is not found.
 	ErrSessionNotFound = errors.New("session not found")
+
+	// ErrDBInitErr is returned when a bucket that we expect to have been
+	// set up during DB initialisation is not found.
+	ErrDBInitErr = errors.New("db did not initialise properly")
 )
 
 // getSessionKey returns the key for a session.
 func getSessionKey(session *Session) []byte {
 	return session.LocalPublicKey.SerializeCompressed()
 }
 
-// StoreSession stores a session in the store. If a session with the
-// same local public key already exists, the existing record is updated/
-// overwritten instead.
-func (db *DB) StoreSession(session *Session) error {
+func (db *DB) CreateSession(session *Session) error {
```

Comment:
Really nice work finding this issue 🚀 🔥!!
Not sure if this is just cause this is a temporary commit and you intend to fix this soon, but I think the `Store` interface also needs to be updated with the `CreateSession` & `UpdateSessionRemotePubKey` functions, and the removal of the `StoreSession` function.
These two new functions also need to have some docs added :)!

Reply:
It's a temp commit meaning that it is just for functionality rn and that it will be cleaned up later as mentioned offline 👍